lib/vector-vrl-functions/src/set_metadata_field.rs

```diff
@@ -101,4 +103,30 @@ impl Expression for SetMetadataFieldFn {
     fn type_def(&self, _: (&state::LocalEnv, &state::ExternalEnv)) -> TypeDef {
         TypeDef::null().infallible()
     }
+
+    fn update_state(
+        &mut self,
+        local: &mut LocalEnv,
+        external: &mut ExternalEnv,
+    ) -> std::result::Result<(), ExpressionError> {
+        if let MetadataKey::Query(query) = &self.key {
+            let new_type = self.value.type_def((local, external)).kind().clone();
+            match query.target() {
+                QueryTarget::External => {
+                    external.update_metadata(
+                        external
+                            .metadata_kind()
+                            .clone()
+                            .with_type_set_at_path(&query.path().to_lookup(), new_type)
+                            // If an unsupported path is used, no type information must be assumed, so any must be returned
+                            .unwrap_or_else(|_| Kind::any()),
+                    );
+                }
+                QueryTarget::Container(_)
+                | QueryTarget::FunctionCall(_)
+                | QueryTarget::Internal(_) => { /* unsupported query types */ }
```

Comment:
Do we want to return an `ExpressionError` here?

Reply:
This branch is unreachable (I updated this to use an actual `unreachable!()`. Only "external" queries are allowed here. In the future, when looks support pointing to metadata this entire function won't be needed, and it will get cleaned up a bit.